### PR TITLE
Connections handler

### PR DESCRIPTION
**Pull request by sstanculeanu**

### CODE REVIEW

File: interface.go [hunk 1]
```diff
@@ -62,15 +61,25 @@ type Messenger interface {
...
+	GetConnectedPeersInfo() *ConnectedPeersInfo
+	WaitForConnections(maxWaitingTime time.Duration, minNumOfPeers uint32)
 	IsConnectedToTheNetwork() bool
 	ThresholdMinConnectedPeers() int
 	SetThresholdMinConnectedPeers(minConnectedPeers int) error
-	SetPeerShardResolver(peerShardResolver PeerShardResolver) error
+	IsInterfaceNil() bool
+}
+
+// Messenger is the main struct used for communication with other peers
```
Comment (ssd04) [inline interface.go:73]: update the comment here? to remove struct

File: libp2p/connectionsHandler.go [hunk 2]
```diff
@@ -0,0 +1,389 @@
...
+func (handler *connectionsHandler) ConnectedPeersOnTopic(topic string) []core.PeerID {
+	return handler.peersOnChannel.ConnectedPeersOnChannel(topic)
+}
+
+// ConnectedFullHistoryPeersOnTopic returns the connected peers on a provided topic
+func (handler *connectionsHandler) ConnectedFullHistoryPeersOnTopic(topic string) []core.PeerID {
+	peerList := handler.ConnectedPeersOnTopic(topic)
+	fullHistoryList := make([]core.PeerID, 0)
+
+	handler.mutPeerResolver.RLock()
```
Comment (iulianpascalau) [inline libp2p/connectionsHandler.go:225]: Good find here for the missing lock. However, I would keep it only to get the `peerShardResolver` pointer

File: libp2p/connectionsHandler.go [hunk 3]
```diff
@@ -0,0 +1,389 @@
...
+		IntraShardObservers:      make(map[uint32][]string),
+		CrossShardValidators:     make(map[uint32][]string),
+		CrossShardObservers:      make(map[uint32][]string),
+		FullHistoryObservers:     make(map[uint32][]string),
+		NumObserversOnShard:      make(map[uint32]int),
+		NumValidatorsOnShard:     make(map[uint32]int),
+		NumPreferredPeersOnShard: make(map[uint32]int),
+	}
+
+	handler.mutPeerResolver.RLock()
```
Comment (iulianpascalau) [inline libp2p/connectionsHandler.go:273]: keep the locking only to get the `peerShardResolver` pointer?

File: libp2p/netMessenger.go [hunk 4]
```diff
@@ -82,21 +81,17 @@ func init() {
 type networkMessenger struct {
 	p2pSigner
 	p2p.MessageHandler
+	p2p.ConnectionsHandler
```
Comment (sstanculeanu) [inline libp2p/netMessenger.go:84]: - move `Bootstrap()` method as well into connectionsHandler
- also move the `printLogsStats` rountine to connectionsHandler, including the `connectionsMetric`

File: libp2p/connectionsHandler.go [hunk 5]
```diff
@@ -0,0 +1,389 @@
...
+			connPeerInfo.NumCrossShardObservers++
+			break
+		}
+
+		connPeerInfo.IntraShardObservers[peerInfo.ShardID] = append(connPeerInfo.IntraShardObservers[peerInfo.ShardID], connString)
+		connPeerInfo.NumIntraShardObservers++
+	}
+}
+
+func (handler *connectionsHandler) peerID() core.PeerID {
```
Comment (iulianpascalau) [inline libp2p/connectionsHandler.go:333]: same question: can we provide directly the peerID on the connectionsHandler's constructor and avoid this cast & function?

File: libp2p/connectionsHandler.go [hunk 6]
```diff
@@ -0,0 +1,389 @@
...
+}
+
+// IsConnectedToTheNetwork returns true if the current node is connected to the network
+func (handler *connectionsHandler) IsConnectedToTheNetwork() bool {
+	netw := handler.p2pHost.Network()
+	return handler.connMonitor.IsConnectedToTheNetwork(netw)
+}
+
+// SetThresholdMinConnectedPeers sets the minimum connected peers before triggering a new reconnection
+func (handler *connectionsHandler) SetThresholdMinConnectedPeers(minConnectedPeers int) error {
```
Comment (iulianpascalau) [inline libp2p/connectionsHandler.go:344]: for the future refactoring: Please consider if this should be part of the connectionMonitor interface & implementation. In this way I think we can remove some circular dependencies

File: libp2p/connectionsHandler.go [hunk 7]
```diff
@@ -0,0 +1,389 @@
...
+
+	for _, p := range handler.p2pHost.Peerstore().Peers() {
+		peers = append(peers, core.PeerID(p))
+	}
+	return peers
+}
+
+// Addresses returns all addresses found in peerstore
+func (handler *connectionsHandler) Addresses() []string {
+	addrs := make([]string, 0)
```
[the remark this answers is not in the document]
Comment (sstanculeanu) [reply libp2p/connectionsHandler.go:92]: right, updated

File: libp2p/connectionsHandler.go [hunk 8]
```diff
@@ -0,0 +1,389 @@
...
+func (handler *connectionsHandler) IsConnected(peerID core.PeerID) bool {
+	h := handler.p2pHost
+
+	connectedness := h.Network().Connectedness(peer.ID(peerID))
+
+	return connectedness == network.Connected
+}
+
+// ConnectedPeers returns the current connected peers list
+func (handler *connectionsHandler) ConnectedPeers() []core.PeerID {
```
[the remark this answers is not in the document]
Comment (sstanculeanu) [reply libp2p/connectionsHandler.go:158]: indeed, thanks

File: libp2p/connectionsHandler.go (hunk 6 above)
Comment (sstanculeanu) [reply libp2p/connectionsHandler.go:344]: added TODO for next PR

**Review by iulianpascalau — APPROVED** [review]

**Review by ssd04 — APPROVED** [review]

**Pull request merged**